java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,351 @@
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than {@code MAX_TAG_LENGTH}) or it's not equal to the prefix of a
+   * correct tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          String.format("The received tag has invalid length. Expected %d or more, got %d.",
+              minPeerTagLength, peerTag.length));
+    }
+
+    byte[] truncatedTag = Arrays.copyOf(generateTag(msg, cnt, this.peerMacKey), peerTag.length);
+    if (!Arrays.equals(peerTag, truncatedTag)) {
+      throw new WrongTagException("The received tag doesn't match the expected tag.");
+    }
+  }
+
+  /**
+   * Common part of the constructors. Initialises peer's public key, minimum length of a peer's tag,
+   * base point for Curve25519. Calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum length of a peer's tag.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  private void initMacKeys(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException, InvalidKeyException {
+    this.peerKey = peerKey;
+    if (minPeerTagLength < MIN_TAG_LENGTH || minPeerTagLength > MAX_TAG_LENGTH) {
+      throw new MinPeerTagLengthOutOfBoundsException(
+          String.format("minPeerTagLength argument should be in [1..32] range. Got %d.",
+              minPeerTagLength));
+    }
+    this.minPeerTagLength = minPeerTagLength;
+
+    KeyAgreement ka;
+    try {
+      ka = KeyAgreement.getInstance("XDH");
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    ka.init(userPrivateKey());
+    ka.doPhase(peerKey, true);
+
+    byte[] sharedSecret = ka.generateSecret();
+    byte[] userPublicKey = subjectPublicKey(userPublicKey());
+    byte[] peerPublicKey = subjectPublicKey(peerKey);
+
+    this.userMacKey = getMacKey(sharedSecret, peerPublicKey, userPublicKey);
+    this.peerMacKey = getMacKey(sharedSecret, userPublicKey, peerPublicKey);
+  }
+
+  /**
+   * Returns a byte representation of a public key.
+   *
+   * @param pk public key.
+   * @return byte representation of {@code pk}.
+   */
+  private byte[] subjectPublicKey(Key pk) {
+    byte[] encodedPK = pk.getEncoded();
+    return Arrays.copyOfRange(encodedPK, encodedPK.length - 32, encodedPK.length);
+  }
+
+  private XECPublicKey derivePublicKeyFromPrivate(XECPrivateKey privateKey) {
+    KeyPairGenerator keyPairGenerator;
+    try {
+      keyPairGenerator = KeyPairGenerator.getInstance("X25519");
```

Comment:
And same point here. But maybe that's going to be obsolete with the migration to Tink.